### ISSUE TITLE
A new fine-tuning method proposal from NeurIPS'22

### ISSUE THREAD

**Issue by zhouyu5** [huggingface/peft#947]
### Feature request

A new parameter-efficient fine-tuning method called `Scaling & Shifting` has been accepted by NeurIPS'22. The paper could be found [here](https://arxiv.org/pdf/2210.08823.pdf), the official code implementation could be found [here](https://github.com/dongzelian/SSF). Do you have time to have a look at this paper and incorporate it in PEFT?

### Motivation

It is tested on CV models, however, after some initial experiment, we found it also applicable to NLP domain. I paste the abstract of the paper as follows.

> Existing fine-tuning methods either tune all parameters of the pre-trained model (full fine-tuning), which is not efficient, or only tune the last linear layer (linear probing), which suffers a significant accuracy drop compared to the full fine-tuning.
In this paper, we propose a new parameter-efficient fine-tuning method termed as SSF, representing that researchers only need to Scale and Shift the deep Features extracted by a pre-trained model to catch up with the performance of full finetuning. In this way, SSF also surprisingly outperforms other parameter-efficient fine-tuning approaches even with a smaller number of tunable parameters. Furthermore, different from some existing parameter-efficient fine-tuning methods (e.g., Adapter or VPT) that introduce the extra parameters and computational cost in the training and inference stages, SSF only adds learnable parameters during the training stage, and these additional parameters can be merged into the original pre-trained model weights via re-parameterization in the inference phase. With the proposed SSF, our model obtains 2.46% (90.72% vs. 88.54%) and 11.48% (73.10% vs. 65.57%) performance improvement on FGVC and VTAB-1k in terms of Top-1 accuracy compared to the full fine-tuning but only fine-tuning about 0.3M parameters. We also conduct amounts of experiments in various model families (CNNs, Transformers, and MLPs) and datasets. Results on 26 image classification datasets in total and 3 robustness & out-of-distribution datasets show the effectiveness of SSF.

### Your contribution

I would like to submit the PR if you think this method could be added to PEFT.

**Comment by BenjaminBossan** [comment]
In general, we consider adding new methods if there is sufficient interested (as documented [here](https://huggingface.co/docs/peft/developer_guides/contributing#adding-a-new-finetuning-method)). It's always a bit subjective to determine the interest, but e.g. in the linked repo, there was only 1 commit since November last year. Maybe other users who are interested could leave a :+1: on the original post to show their interest.

Regarding the method itself, at first glance the scale-part looks similar (though not identical) to the IA³-part of [this paper](https://arxiv.org/abs/2205.05638), which is already implemented in PEFT. Maybe it would be more efficient to add the shift-part to IA³ then to add a completely new method.

**Comment by zhouyu5** [comment]
> In general, we consider adding new methods if there is sufficient interested (as documented [here](https://huggingface.co/docs/peft/developer_guides/contributing#adding-a-new-finetuning-method)). It's always a bit subjective to determine the interest, but e.g. in the linked repo, there was only 1 commit since November last year. Maybe other users who are interested could leave a 👍 on the original post to show their interest.
> 
> Regarding the method itself, at first glance the scale-part looks similar (though not identical) to the IA³-part of [this paper](https://arxiv.org/abs/2205.05638), which is already implemented in PEFT. Maybe it would be more efficient to add the shift-part to IA³ then to add a completely new method.

Thanks for your valuable reply, I will have a look at the IA3 paper before I go into the details with you.

**Comment by zhouyu5** [comment]
Hi @BenjaminBossan 

Thanks again for your valuable reply.

Before submitting this issue, I have already read through carefully the whole [contribution guide](https://huggingface.co/docs/peft/developer_guides/contributing#adding-a-new-finetuning-method), as you said, determining whether a method is '_sufficient interested_' is '_a bit subjective_'. I totally understand your caution before adding a new method, however, I slightly disagree with your use of votes or the number of GitHub commits as an indicator of whether a new approach is interesting. The reason for my disagreement is as follows:

- In the contribution guide, it does not clearly state how many submissions are required for the corresponding GitHub repository.
- On the other hand, If I understand correctly, the key principle of adding a new method is to prove its value instead of its popularity (although sometimes popularity can reflect value). I think we both agree that we should provide enough evidence instead of enough votes.

Regarding your other proposal, I read the $IA^3$ paper carefully, and I agree with your insights, they are similar methods, and their scale part is almost identical. I also proposed adding the shift-part to the $IA^3$ method for the following reasons:

- According to the ablation study of the `Scaling & Shifting` paper, the shift part also contributes to the final performance, I paste the performance data here (see the highlighted part):
<img width="159" alt="image" src="https://github.com/huggingface/peft/assets/12825276/84fad7e9-16ac-4c45-b917-4c48b5af9375">

Besides, I want to emphasize the reason why I want to add the method,

1. There is no single method that can ensure the best results on all data sets and tasks. I believe that by improving the coverage of the algorithm and combining some algorithm attempts, we can better help users find the best algorithm for their tasks.
2. Also, I found the `Scaling & Shifting` is a good supplement to the PEFT library, since it covers the CV domain, and therefore help PEFT demonstrate the universality and effectiveness of various domain (not only the NLP domain but also CV domain)

For the next steps, I would like to make it clear to you:

1. Implement the shift-part in $IA^3$ method, let's call it advanced $IA^3$ method
2. Invoke the advanced $IA^3$ method as the `Scaling & Shifting` method, add some examples, and test scripts

Please check it and expect your invaluable feedback! 

Thank you!

**Comment by BenjaminBossan** [comment]
Thanks for taking a closer look and coming up with a proposal.

I agree that thumbs up counts are not a perfect measurement, but there probably isn't any single best measurement. 

Let me explain the reasons for being cautious: For us maintainers, adding a new method always comes with the trade-off of maintenance burden. It can often happen that the original contributors of a new method will not help when users encounter issues with the method or when it needs to be upgraded/refactored, so all the burden will potentially fall on us. At the same time, papers often claim great results that fail to replicate in practice, so community adoption is a good (but imperfect) indicator for the robustness of results.

That being said, if the proposed method can be added on top of IA³ with only a few additions, the cost would be minimal and we can certainly accept less evidence for the method's utility than if it required hundreds of new lines of code.

@younesbelkada @pacman100 do you have an opinion?

**Comment by zhouyu5** [comment]
Hi @BenjaminBossan 

Thanks for your explanation of being cautious. It makes sense to me. I agree with you that there isn't any single best measurement. Many thanks to you for taking my proposal of adding a new method into consideration. Besides, I enjoy talking to you, which greatly helps me. I am looking forward to further discussion with you guys.

**Comment by zhouyu5** [comment]
Hello, and updates on that?
@BenjaminBossan @younesbelkada @pacman100

**Comment by zhouyu5** [comment]
Hello, and updates on that?
Will it be ok for me to submit a PR?
